File: Week 2/python/1.data_types_review.py
"""
WEEK 2 - PYTHON CONTROL FLOW
Topic: Data Types Review
File: 1.data_types_review.py

CONCEPT:
A quick refresher of the Java examples in this lesson:
  1. Byte overflow (Java only) — Python has no byte/short/int distinction.
  2. Integer-division pitfall when storing 6/4 in a float.
  3. Modulo on float vs int.
  4. Relational operator returns bool.

Python's behaviour differs from Java in subtle but important ways:
  - `/` is float division, `//` is integer (floor) division.
  - int is arbitrary precision, so there is no overflow.
  - bool is a subtype of int; True == 1, False == 0.

KEY POINTS:
 - 6 / 4   -> 1.5    (Java would give 1; use 6 // 4 to match Java's int /).
 - 55.5 % 10 -> 5.5  (modulo works on floats).
 - 5 > 6    -> False (Python prints the literal `False`).
 - No fixed-width byte type — but Python WOULD let you build one via bytes().

SYNTAX:
 a / b              # float division
 a // b              # floor division
 a % b              # modulo (sign follows divisor!)
 a > b              # -> bool

DRY RUN:
 1) a = 6 / 4         -> 1.5  (float)
    b = 6 // 4         -> 1    (int)
    c = a + b           -> 2.5  (mixed -> float)
 2) 55.5 % 10           -> 5.5
    55   % 10           -> 5
 3) 5 > 6                -> False

COMPLEXITY: O(1) for each example.
"""


def example_1_no_byte_overflow() -> None:
    """Python has no `byte` type, so the Java overflow case does not apply."""
    print("--- Example 1: byte overflow (Java only) ---")
    # In Java: byte b = 50 * 50;  // 2500 doesn't fit in byte; compile error.
    # In Python: int is arbitrary-precision, so 50 * 50 = 2500 just works.
    b = 50 * 50
    print(f"50 * 50 = {b}   # Python int has no fixed width")


def example_2_division() -> None:
    """6/4 in Python differs from Java -- here we get FLOAT, not INT."""
    print("\n--- Example 2: division ---")
    a: float = 6 / 4       # 1.5 (true division)
    b: int = 6 // 4         # 1   (floor division -- matches Java's int /)
    c: float = a + b        # 2.5
    print(f"a (6/4)  = {a}")
    print(f"b (6//4) = {b}")
    print(f"c = a + b = {c}")


def example_3_modulo() -> None:
    """Modulo works for both int and float."""
    print("\n--- Example 3: modulo on float vs int ---")
    a = 55.5
    b = 55
    a = a % 10              # 5.5
    b = b % 10              # 5
    print(f"55.5 % 10 = {a}")
    print(f"55   % 10 = {b}")


def example_4_relational() -> None:
    """A relational operator returns a bool."""
    print("\n--- Example 4: relational returns bool ---")
    var1, var2 = 5, 6
    print(f"5 > 6 -> {var1 > var2}")    # False


def main() -> None:
    example_1_no_byte_overflow()
    example_2_division()
    example_3_modulo()
    example_4_relational()


if __name__ == "__main__":
    main()


# NOTES:
# - Java's primitive integer types overflow silently (or in narrow types, cause a
#   compile error). Python's `int` is unbounded.
# - Java `/` truncates ints, `%` follows the SIGN OF THE DIVIDEND.
#   Python `/` is float division, `%` follows the SIGN OF THE DIVISOR.
# - Bool is a subclass of int in Python; `True + 1 == 2` is legal (Java forbids it).

File: Week 2/cpp/1.data_types_review.cpp
/*
 * WEEK 2 - C++ CONTROL FLOW
 * Topic: Data Types Review
 * File: 1.data_types_review.cpp
 *
 * CONCEPT:
 * Re-examine type behaviour with arithmetic and relational operators.
 *  1. char/byte overflow vs int promotion.
 *  2. Integer-division pitfall when assigning to double.
 *  3. Modulo on int (and float via fmod).
 *  4. Relational operator returns bool.
 *
 * KEY POINTS:
 *  - C++ implicitly promotes char/short to int during arithmetic.
 *  - 6 / 4 == 1 (int division); store in a double -> 1.0 (no precision recovered).
 *  - `%` works on integers; for floats use `std::fmod`.
 *  - Relational operators return bool (true/false).
 *
 * SYNTAX:
 *   double a = 6 / 4;          // 1 (then converted to 1.0)
 *   double b = 6.0 / 4;        // 1.5 (mixed-type promotion)
 *   std::fmod(55.5, 10.0);     // 5.5
 *
 * DRY RUN:
 *   a = (double)(6/4) -> 1.0
 *   b = 6/4 -> 1
 *   c = a + b = 2.0
 *   fmod(55.5, 10) = 5.5
 *   (5 > 6) -> false
 *
 * COMPLEXITY: O(1).
 */

#include <iostream>
#include <cmath>
#include <cstdint>

int main() {
    std::cout << std::boolalpha;

    // Example 1: byte-like overflow
    // In C++ a `signed char` is similar to Java's byte. Assigning beyond [-128,127]
    // would silently wrap (implementation-defined). The compiler doesn't reject it.
    std::int8_t sb = 50;
    int product = sb * 50;        // sb is PROMOTED to int -> 2500 fits
    std::cout << "int8_t 50 * 50 = " << product
              << "  (promotion saves the day)\n";

    // Example 2: integer-division pitfall
    double a = 6 / 4;             // 1, then 1.0
    int    b = 6 / 4;             // 1
    double c = a + b;             // 2.0
    std::cout << "(double)(6/4) + (int)(6/4) = " << c << '\n';

    // Example 3: modulo on int vs fmod for floats
    double x = 55.5;
    int    y = 55;
    std::cout << "55.5 fmod 10 = " << std::fmod(x, 10.0) << '\n';   // 5.5
    std::cout << "55  %   10  = " << (y % 10)          << '\n';     // 5

    // Example 4: relational operators return bool
    int v1 = 5, v2 = 6;
    std::cout << "(5 > 6) = " << (v1 > v2) << '\n';                 // false

    return 0;
}

/*
 * NOTES:
 *  - Java auto-rejects narrow-type overflows at compile time for known
 *    constants; C++ silently wraps signed integers (UB) -- be careful.
 *  - `%` operates ONLY on integers in C++. Use `std::fmod` (from <cmath>) for floats.
 *  - `std::boolalpha` makes cout print `true`/`false` instead of 1/0.
 */
